### DISCUSSION TITLE
Better way to create a seccomp filter for a specific language

### DISCUSSION THREAD

**Discussion by AlphaNecron** [criyle/go-judge#77] [Q&A]
![image](https://github.com/criyle/go-judge/assets/57827456/7e9fecb5-0052-41d0-ac3f-5a8bda2bfd1f)
Is there a faster and painless way to generate seccomp configs for languages besides manually adding them? Btw should I use container or ptrace, which is better though?

**Comment by AlphaNecron** [comment]
Anyways, switched to Linux namespace. I'm doing a worker, the app will run indefinitely and take submissions from queue, each submission has `x` test cases, and for each test case do I have to spawn a new `ns` and `cgroup` instance or use a shared instance for everything?

**Reply by criyle** [reply]
If you use `ns` worker type, a new namespace will be created for every request. If you use `container`, the namespace will be created with an `init` process and being reused. However, cgroup is being created for every request for better tracking memory and cpu usages.

**Reply by AlphaNecron** [reply]
Oh, thanks. So creating a new `ns` instance for every submission is the only way right?

**Reply by AlphaNecron** [reply]
Anyways, with `container`, how to specify seccomp policy for each run? Is it possible?

**Reply by criyle** [reply]
It is not possible yet, PRs welcome.

**Reply by AlphaNecron** [reply]
Oh, so we have to use a unified seccomp policy for everything? Is it safe to do that?

**Reply by AlphaNecron** [reply]
Anyways, for each core, should I spawn a container all jobs or a container pool for each core?

**Reply by AlphaNecron** [reply]
Btw, does `cgroup` package support nested cgroup? Using nested cgroup makes it easier to monitor stuff like CPU and mem usage.

**Reply by AlphaNecron** [reply]
Can we push a file to the container directly to compile it?

**Reply by AlphaNecron** [reply]
> Can we push a file to the container directly to compile it?

Can I do environment.Open() to get a File object and write the content to that file?

**Reply by AlphaNecron** [reply]
Nice, this approach works flawlessly. However, compiler throws `vfork: function not implemented` with `vfork` added to seccomp whitelist. Are there security concerns with seccomp disabled? I'm using container.

**Reply by criyle** [reply]
> Anyways, for each core, should I spawn a container all jobs or a container pool for each core?

The `executor_server` uses container pool for each core

>Btw, does cgroup package support nested cgroup? Using nested cgroup makes it easier to monitor stuff like CPU and mem usage.

Not yet, you can set the prefix of the cgroup to a existing cgroup so that you can control the total resource usage in that way.

>Can we push a file to the container directly to compile it?

If you are using the `executor_server`, you can specify `copyIn` parameter to copy a file into the container before the execution.

>Can I do environment.Open() to get a File object and write the content to that file?

Yes, that is how `executor_server` copies file into the container [here](https://github.com/criyle/go-judge/blob/master/envexec/file_copyin.go#L50-L62)

**Reply by AlphaNecron** [reply]
> The `executor_server` uses container pool for each core

How does that work, it binds to a core using cgroup or something? Btw, I'm trying to get the compiler output, I'm doing `memfd.New` and pass `fd` to Files property; however, that does not seem to work.
Here's my implementation
![image](https://github.com/criyle/go-judge/assets/57827456/58cbac99-b882-4f42-b89b-8b376a1dcc43)
![image](https://github.com/criyle/go-judge/assets/57827456/c5c22bd1-2186-4b23-9324-b4e0f93b3022)

The code is kinda long so I'm posting entrypoints only. `compOut` should store compiler output, but it is empty after running `g++` with `--verbose`.

**Reply by AlphaNecron** [reply]
Dumping to `stdout` and `stderr` is fine.

**Reply by criyle** [reply]
File descriptor is being duplicated into child processes so the file pointer is changed by writing as well. You may use `mf.Seek(0, 0)` to reset the file pointer before reading the file.

**Reply by AlphaNecron** [reply]
Thanks.

**Reply by AlphaNecron** [reply]
How does `container` work? It uses `containerd` under the hood?

**Reply by criyle** [reply]
Nope, it uses kernel container interface directly in [this package](https://github.com/criyle/go-sandbox).

**Reply by AlphaNecron** [reply]
Oh, thanks.
